src/lib/components/TableOfContents/index.js

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {html} from 'lit-element';
+import {BaseStateElement} from '../BaseStateElement';
+import {closeToC} from '../../actions';
+
+/**
+ * Element that renders table of contents.
+ * @extends {BaseStateElement}
+ * @final
+ */
+class TableOfContents extends BaseStateElement {
+  static get properties() {
+    return {
+      title: {type: String},
+      titleId: {type: String, attribute: 'title-id'},
+      opened: {type: Boolean, reflect: true},
+    };
+  }
+
+  constructor() {
+    super();
+    this.scrollSpy = this.scrollSpy.bind(this);
+    this.openedToFalse = this.openedToFalse.bind(this);
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.tocHTML = this.innerHTML;
+    this.divContent =
+      document.querySelector('#content') || document.createElement('div');
+    this.headings = this.divContent.querySelectorAll('h1[id], h2[id], h3[id]');
+
+    this.observer = new IntersectionObserver(this.scrollSpy);
+    this.headings.forEach((heading) => this.observer.observe(heading));
+  }
+
+  render() {
+    return html`
+      <div class="w-toc__label">
+        <span>In this article</span>
+        <button
+          class="w-button w-button--secondary w-button--icon"
+          data-icon="close"
+          aria-close="Close Table of Contents"
+          @click="${this.openedToFalse}"
+        ></button>
+      </div>
+      <div class="w-toc__content">
+        <h2 class="w-toc__header">
+          <a href="#${this.titleId}" class="w-toc__header--link">
+            ${this.title}
+          </a>
+        </h2>
+        ${html([this.tocHTML])}
+      </div>
+    `;
+  }
+
+  updated(changedProperties) {
+    if (
+      changedProperties.has('opened') &&
+      changedProperties.get('opened') !== this.opened
```

Comment:
So the diff check between the previous state and the new state is necessary for when you open/close a the toc then go to another page. If I just checked `changedProperties.get('opened')` (the previous value) it would falsely trigger the `#content` to have to `.w-toc-open` but would not update the property of the element. It was weird, so this seems to work well (I'm not 100% why it behaves this way)